### DISCUSSION TITLE
Is it normal to do an API request inside getServerSideProps?

### DISCUSSION THREAD

**Discussion by hitrop** [vercel/next.js#25445] [Help]
Currently, I have my backend in NodeJS (Express) and also currently in process of migrating from create-react-app to NextJS on the frontend. Should I do API requests to my express server from getServerSideProps OR should I move the API logic (db query, etc) to the getServerSideProps itself (with NextJS interacting with the database directly)? 

Is it normal to write the db and entire API logic in NextJS itself? Or should I go with the more traditional approach of NextJS (getServerSideProps) interacting with my express server through fetch API calls?

I understand that writing the API logic in NextJS itself will be faster in terms of page loading speed since it would now remove the additional fetch calls inside getServerSideProps. But this would strongly couple my frontend and backend logic. Also, I would have to handle the session thing in NextJS (in addition to doing it in express).

So, my question is which of the two approaches is better in long run ():

1. NextJS interacting with APIs written in a separate backend server (Express server)

2. Move some/all of the backend logic to NextJS (getServerSideProps, with NextJS directly interacting with Database)

Also, assuming the bandwidth of the server (where the NextJS app is deployed) is good enough, the additional fetch calls inside getServerSideProps shouldn't be an issue, right? Or am I missing something here?

**Comment by JacobLey** [comment]
tldr: Use your own custom API server and `fetch` responses in your NextJS app. (option 1)

My reasoning:

1) Checkout the "compiled" version of your code when building it. NextJS creates a completely different copy of your code for every page. So if every page uses a shared version of some database connector, that gets copied N times. Which can be means pages are not actually sharing a connection, and must be initialized individually. If you are deploying on Vercel, their internal usage of lambda functions might have weird effects on connections/data that is supposed to be long lived.
2) It sounds like you are primarily concerned about bandwidth from your NextJS app. Note that if you are connecting to any external data source (database, storage, api, etc...) you will incur bandwidth no matter what. It is just a difference of where that bandwidth is going. Depending on your database queries, you may actually incur less bandwidth by only connecting to the API.
3) If all that is happening in `getServerSideProps` is API calls, you could even move that logic to `getInitialProps`. The benefit there is that you would only need to make API calls if the response is not already cached client side. The extra cache check might be more complicated than you care about for now, but you _could_ implement it later if you want. No API calls is way faster than server-side api calls!
4) Using a custom server gives you _way_ more versatility. NextJS itself comes with some custom API logic, but your Express server is going to be much more capable than anything NextJS has. You aren't even restricted to using Javascript if you want, so your frontend can be NextJS, and backend can be something else like Python or Go.
5) Do one thing, and do it right. NextJS is great for providing server-side rendering of frontend code. That improves SEO and user experience. NextJS was not built for all the business/database logic that should exist only on the backend. Decouple the two things for easiest development. As your app _really_ grows, you may start to use services like message queues or scheduled jobs which don't fit well in the NextJS framework.

In fairness, I'll give some counter arguments:

1) Doing everything in NextJS is easier. If you aren't building some big app, but rather some simplistic landing page, blog, or personal project, then it is fine to keep in all in NextJS. Odds are your custom server is trivial and NextJS should be able to support all of that functionality.
2) NextJS explicitly tells you to do this: https://nextjs.org/docs/basic-features/data-fetching I would generally argue this is disingenuous of NextJS. As stated above, if your goal is something simple then sure keep it in one place. But mainly NextJS is sponsored by Vercel, whose long term goal is to get you to use their platform for your entire codebase, and pay them while you do it. While deploying with Vercel is still fine, I wouldn't sacrifice quality of code or performance for a private companies goals.

So with all that in mind, I would definitely encourage custom server all the way.
So unless you understand (and accept) the limitations of pure NextJS, I wouldn't recommend that do that
